src/clean-core/allocator.cc

```diff
@@ -0,0 +1,199 @@
+#include <clean-core/allocator.hh>
+
+#include <cstdlib>
+#include <cstring>
+
+#include <clean-core/assert.hh>
+#include <clean-core/utility.hh>
+
+namespace
+{
+struct stack_alloc_header
+{
+    size_t padding;
+};
+
+template <class T>
+[[nodiscard]] T align_up_masked(T value, size_t mask)
+{
+    return (T)(((size_t)value + mask) & ~mask);
+}
+
+template <class T>
+[[nodiscard]] T align_up(T value, size_t alignment)
+{
+    return align_up_masked(value, alignment - 1);
+}
+
+inline std::byte* align_up_with_header(std::byte* head, size_t align, size_t header_size)
+{
+    std::byte* padded_res = align_up(head, align);
+    size_t const padding = size_t(padded_res - head);
+
+    if (padding < header_size)
+    {
+        // header does not fit - align up
+        header_size -= padding;
+
+        if (header_size % align > 0)
+        {
+            padded_res += align * (1 + (header_size / align));
+        }
+        else
+        {
+            padded_res += align * (header_size / align);
+        }
+    }
+
+    return padded_res;
+}
+}
+
+cc::byte* cc::linear_allocator::alloc(cc::size_t size, cc::size_t align)
+{
+    CC_ASSERT(_buffer != nullptr && "linear_allocator uninitialized");
+
+    auto* const padded_res = align_up(_head, align);
+
+    if (CC_UNLIKELY(padded_res + size > _buffer_end))
+    {
+        CC_ASSERT(false && "linear_allocator overcommited");
+        return nullptr;
```

Comment:
this could be replaced by CC_UNREACHABLE

Reply:
done